IO/Storage: CPU-bound and cross-thread scheduling

kprotty opened the pull request:
This introduces the concept of an `intent` enum for `Storage.on_next_tick` which controls how the NextTick is scheduled. The enum includes: `yield` which defaults to `IO.timeout(0)` as usual, `cpu_work` which runs the NextTick on a thread pool, and `cpu_inject` which brings the NextTick back from a random (pool) thread to the IO thread. This is done with `tb_client.Signal` which now uses proper IO events (eventfd, EVFILT_USER, PostQueuedCompletionStatus) instead of a hacky, local TCP server.

The following LSM bits now utilize `cpu_work` to offload heavy tasks from the main thread:
* Sorting of `TableMutable` to `TableImmutable`.
* `TableBuilder.data_block_finish` in `Compaction` (this calls `vsr.calculate_checksum_body`)
* `Grid.read_block_validate` after a `Storage.read_sectors` occurs (this also calls `vsr.calculate_checksum_body`).

## Pre-merge checklist

Performance:

* [x] Compare `zig benchmark` on linux before and after this pr.
    ``` sh
    # benchmark results before
    1223 batches in 135.30 s
    load offered = 1000000 tx/s
    load accepted = 73908 tx/s
    batch latency p00 = 0 ms
    batch latency p10 = 28 ms
    batch latency p20 = 29 ms
    batch latency p30 = 30 ms
    batch latency p40 = 30 ms
    batch latency p50 = 31 ms
    batch latency p60 = 32 ms
    batch latency p70 = 33 ms
    batch latency p80 = 34 ms
    batch latency p90 = 36 ms
    batch latency p100 = 5517 ms
    transfer latency p00 = 0 ms
    transfer latency p10 = 4166 ms
    transfer latency p20 = 10805 ms
    transfer latency p30 = 21106 ms
    transfer latency p40 = 34218 ms
    transfer latency p50 = 49408 ms
    transfer latency p60 = 64719 ms
    transfer latency p70 = 74813 ms
    transfer latency p80 = 90404 ms
    transfer latency p90 = 106386 ms
    transfer latency p100 = 125316 ms
    
    # benchmark results after
    1223 batches in 82.40 s
    load offered = 1000000 tx/s
    load accepted = 121362 tx/s
    batch latency p00 = 1 ms
    batch latency p10 = 27 ms
    batch latency p20 = 28 ms
    batch latency p30 = 29 ms
    batch latency p40 = 30 ms
    batch latency p50 = 32 ms
    batch latency p60 = 33 ms
    batch latency p70 = 34 ms
    batch latency p80 = 35 ms
    batch latency p90 = 37 ms
    batch latency p100 = 3281 ms
    transfer latency p00 = 1 ms
    transfer latency p10 = 2101 ms
    transfer latency p20 = 5877 ms
    transfer latency p30 = 11111 ms
    transfer latency p40 = 17961 ms
    transfer latency p50 = 26298 ms
    transfer latency p60 = 35204 ms
    transfer latency p70 = 44132 ms
    transfer latency p80 = 53201 ms
    transfer latency p90 = 62509 ms
    transfer latency p100 = 72410 ms
    ```

kprotty committed:
storage: NextTick calls directly into IO

Since `io.timeout(0)` no longer has to go through syscalls, we can
just call it directly instead of buffering NextTick nodes in Storage.

kprotty committed:
storage: Introduce intents for `on_next_tick`

Piggy-backing on the API for scheduling arbitrary work,
NextTick nodes now require their intent to be specified.
This will allow for specializing on how work is scheduled/executed
with the new options being `cpu_work`to defer compute heavy tasks
with a thread pool and `cpu_inject` to get control flow back to the main thread.

kprotty committed:
lsm/storage: cpu_work/cpu_inject as ThreadPool across LSM

kprotty committed:
io: add Events for efficient cross-thread notification

This supersedes the tri-socket, local TCP server hack being
used in `tb_client.Signal` to facilitate cross-thread notification.
Instead, it uses the most efficient mechanicsm for waking the IO
abstraction: `eventfd` on linux, `EVFILT_USER` on macos, and
`PostQueuedCompletionStatus` on windows.

kprotty committed:
io: fix open_event() ErrorSet

Needed to match that of `tb_client.Context`.

cb22 commented on src/clients/c/tb_client/signal.zig:
```diff
@@ -13,274 +13,73 @@ const log = std.log.scoped(.tb_client_signal);
 /// It does this by using OS sockets (which are thread safe) 
```
Believe this comment is out of date now?

jamii commented on src/io/linux.zig:
```diff
@@ -960,6 +960,84 @@ pub const IO = struct {
+
+        self.read(
+            *anyopaque,
+            @intToPtr(*anyopaque, event),
```
Why does io.read take a type argument if the only type that works is `*anyopaque`?

kprotty replied:
It only supports a pointer type, hence the use of `anyopaque` just for representing the `eventfd` as pointer.

jamii replied:
Oh, I saw that it doesn't cast `context` [here](https://github.com/tigerbeetledb/tigerbeetle/blob/0b15d2209f6d51f2f56e7f7a597540b78ddacdf6//src/io/linux.zig#L652), but I didn't know that there is an implicit conversion from *T to *anyopaque.

jamii commented on src/io/linux.zig:
```diff
@@ -960,6 +960,84 @@ pub const IO = struct {
+        errdefer os.close(event_fd);
+
+        // Start listening for notifications on this event.
+        const event = @intCast(u32, event_fd);
```
Why cast this from `fd_t`?

kprotty replied:
Wanted a unified "event" type across impls so went with `u32`. I can go with i32 so fd_t is compatible.

jamii replied:
I guess the cast can't fail because the stdlib already checked for errors?

kprotty replied:
Yea, although I've changed it to be i32 now for compatibility.

jamii commented on src/storage.zig:
```diff
@@ -69,77 +71,178 @@ pub const Storage = struct {
+        notified: bool = false,
+        idle: u16 = 0,
+
+        var instance = ThreadPool{};
```
Why does this need to be global, rather than a field on storage?

kprotty replied:
Good note. Accidentally carried over from when multiple storages was up in the air and Storage didn't support intrusive initialization.

jamii commented on src/storage.zig:
```diff
@@ -69,77 +71,178 @@ pub const Storage = struct {
+
     io: *IO,
     fd: os.fd_t,
+    main_thread_id: std.Thread.Id,
```
Can you use this for `assert(storage.is_background_thread())` in callbacks that are run in the threadpool (eg `cpu_sort_callback`), to call out that they need to be read carefully.

kprotty replied:
This would need to be exposed/work on `testing/storage.zig` too (as Tree is used on both Storage types). But test storage doesn't use threads.

jamii replied:
Good point. We could have testing/storage store a boolean flag that's true when it's executing a background callback?

Just thinking ahead to when I inevitably mess up thread safety :)

jamii commented on src/storage.zig:
```diff
@@ -69,77 +71,178 @@ pub const Storage = struct {
 
+    pub const NextTickIntent = enum {
+        yield,
+        cpu_work,
```
`yield/work/inject` is not super obvious to me. Maybe `same_thread/move_to_foreground_thread/move_to_background_thread`?

jamii replied:
Similarly `is_local` doesn't connect to `work/inject`. Maybe `is_foreground_thread`.

kprotty replied:
Thoughts on `main_thread` instead of `foreground_thread`?

jamii replied:
Sure :)

jamii commented on src/storage.zig:
```diff
@@ -69,77 +71,178 @@ pub const Storage = struct {
+        cpu_inject,
+    };
+
     pub fn on_next_tick(
```
In the doc comment here, could you explain the rules we have to follow for thread safety.

jamii commented on src/testing/storage.zig:
```diff
@@ -275,11 +275,24 @@ pub const Storage = struct {
         next_tick: *Storage.NextTick,
+        intent: NextTickIntent,
     ) void {
+        // All on_next_tick requests are fullened through next_tick_queue for determinism.
```
'fulfilled'?

jamii committed:
Trace thread pool callbacks

jamii commented:
@kprotty I added some tracing. 

![image](https://user-images.githubusercontent.com/340884/231241837-3b596daf-b7f1-4d7b-9884-67c965e78dee.png)

Utilization is kind of low. I wonder if that's due to the control plane not supplying enough work, or to the overhead of communication.

jamii commented:
`tracer` is not thread-safe because of https://github.com/tigerbeetledb/tigerbeetle/blob/6e4ed3bacdb865f1b71979a26ac601cdce7c7875//src/tracer.zig#L337. Since the yield and thread-pool queue are traced, this can cause UB and crashes when tracing benchmarks.

Making the interns and message_buffer thread-local might be sufficient? Tracy itself is thread-safe.

kprotty committed:
storage: ThreadPool local to storage + cleanup

kprotty committed:
io: use i32 for Event

kprotty committed:
storage: introduce ExecutionContext

Tasks in the storage run under a given execution context.
Those currently supported are `main_thread` and `background_thread`.
Users of storage can call `.context()` to query which one they're currently in.
`on_next_tick` also now takes a target context to execute in instead of a "purpose".

kprotty commented:
We decided to focus on improving single threaded performance first. There's still a few more things to implement there.
